File: frontend/src/api/audioJournal.ts
import { buildApiUrl, parseJsonResponse } from "./client";

export type AudioJournalTakeRecord = {
  id: number;
  entry_id: number;
  take_number: number;
  take_type: string;
  created_at: string;
  audio_path: string;
  audio_filename: string;
  transcript_text: string | null;
  transcript_source: string;
  transcription_status: string;
  transcription_engine: string | null;
  transcription_model: string | null;
  duration_seconds: number | null;
  sample_rate: number | null;
  channels: number | null;
  file_format: string | null;
  quality_status: string;
  quality_score: number | null;
  quality_summary: string | null;
  quality_reasons_json: string | null;
  noise_floor_db: number | null;
  rms_db: number | null;
  peak_db: number | null;
  clipping_detected: boolean;
  silence_ratio: number | null;
  snr_estimate_db: number | null;
  is_active: boolean;
  is_training_candidate: boolean;
  training_quality: string | null;
  script_match_score: number | null;
  metadata_json: string | null;
};

export type AudioJournalEntryRecord = {
  id: number;
  title: string;
  created_at: string;
  updated_at: string;
  journal_date: string;
  script_text: string | null;
  original_transcript_text: string | null;
  notes: string | null;
  tags_json: string | null;
  voice_style: string | null;
  active_take_id: number | null;
  selected_training_take_id: number | null;
  overall_quality_status: string;
  metadata_json: string | null;
  takes: AudioJournalTakeRecord[];
};

export type AudioQualityBaselineRecord = {
  id: number;
  name: string;
  created_at: string;
  updated_at: string;
  source_audio_path: string;
  source_audio_filename: string;
  notes: string | null;
  device_label: string | null;
  environment_label: string | null;
  sample_rate: number | null;
  channels: number | null;
  duration_seconds: number | null;
  file_format: string | null;
  peak_db: number | null;
  rms_db: number | null;
  noise_floor_db: number | null;
  snr_estimate_db: number | null;
  silence_ratio: number | null;
  clipping_detected: boolean;
  quality_score: number | null;
  is_default: boolean;
  metadata_json: string | null;
};

export type AudioQualityBaselineListResponse = {
  items: AudioQualityBaselineRecord[];
};

export type AudioQualityBaselineUpdatePayload = Partial<{
  name: string;
  notes: string | null;
  device_label: string | null;
  environment_label: string | null;
  is_default: boolean;
  metadata_json: string | null;
}>;

export type AudioJournalEntryUpdatePayload = Partial<{
  title: string;
  journal_date: string;
  script_text: string | null;
  original_transcript_text: string | null;
  notes: string | null;
  tags_json: string | null;
  voice_style: string | null;
  active_take_id: number | null;
  selected_training_take_id: number | null;
  overall_quality_status: string;
  metadata_json: string | null;
}>;

export type AudioJournalTakeUpdatePayload = Partial<{
  transcript_text: string | null;
  transcript_source: string | null;
  transcription_status: string | null;
  transcription_engine: string | null;
  transcription_model: string | null;
  quality_status: string | null;
  quality_score: number | null;
  quality_summary: string | null;
  quality_reasons_json: string | null;
  is_training_candidate: boolean | null;
  training_quality: string | null;
  script_match_score: number | null;
  metadata_json: string | null;
}>;

export type AudioJournalTrainingCandidatePayload = {
  is_training_candidate: boolean;
  manual_override?: boolean;
  reason?: string | null;
};

export type AudioJournalListResponse = {
  items: AudioJournalEntryRecord[];
};

export type AudioJournalUploadResponse = {
  entry: AudioJournalEntryRecord;
  take: AudioJournalTakeRecord;
};

export async function listAudioJournalEntries() {
  const response = await fetch(await buildApiUrl("/api/audio-journal"));
  return parseJsonResponse<AudioJournalListResponse>(response);
}

export async function getAudioJournalEntry(entryId: number) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/${entryId}`));
  return parseJsonResponse<AudioJournalEntryRecord>(response);
}

export async function listAudioQualityBaselines() {
  const response = await fetch(await buildApiUrl("/api/audio-journal/baselines"));
  return parseJsonResponse<AudioQualityBaselineListResponse>(response);
}

export async function createAudioQualityBaseline(formData: FormData) {
  const response = await fetch(await buildApiUrl("/api/audio-journal/baselines"), {
    method: "POST",
    body: formData,
  });

  return parseJsonResponse<AudioQualityBaselineRecord>(response);
}

export async function updateAudioQualityBaseline(baselineId: number, payload: AudioQualityBaselineUpdatePayload) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/baselines/${baselineId}`), {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  return parseJsonResponse<AudioQualityBaselineRecord>(response);
}

export async function deleteAudioQualityBaseline(baselineId: number) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/baselines/${baselineId}?delete_audio=true`), {
    method: "DELETE",
  });

  if (!response.ok) {
    await parseJsonResponse(response);
  }
}

export async function setDefaultAudioQualityBaseline(baselineId: number) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/baselines/${baselineId}/set-default`), {
    method: "POST",
  });

  return parseJsonResponse<AudioQualityBaselineRecord>(response);
}

export async function buildAudioQualityBaselineAudioUrl(baselineId: number) {
  return buildApiUrl(`/api/audio-journal/baselines/${baselineId}/audio`);
}

export async function createAudioJournalEntry(formData: FormData) {
  const response = await fetch(await buildApiUrl("/api/audio-journal"), {
    method: "POST",
    body: formData,
  });

  return parseJsonResponse<AudioJournalUploadResponse>(response);
}

export async function updateAudioJournalEntry(entryId: number, payload: AudioJournalEntryUpdatePayload) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/${entryId}`), {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  return parseJsonResponse<AudioJournalEntryRecord>(response);
}

export async function deleteAudioJournalEntry(entryId: number) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/${entryId}`), {
    method: "DELETE",
  });

  if (!response.ok) {
    await parseJsonResponse(response);
  }
}

export async function listAudioJournalTakes(entryId: number) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/${entryId}/takes`));
  return parseJsonResponse<AudioJournalTakeRecord[]>(response);
}

export async function createAudioJournalTake(entryId: number, formData: FormData) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/${entryId}/takes`), {
    method: "POST",
    body: formData,
  });

  return parseJsonResponse<AudioJournalTakeRecord>(response);
}

export async function updateAudioJournalTake(
  entryId: number,
  takeId: number,
  payload: AudioJournalTakeUpdatePayload,
) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/${entryId}/takes/${takeId}`), {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  return parseJsonResponse<AudioJournalTakeRecord>(response);
}

export async function deleteAudioJournalTake(entryId: number, takeId: number) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/${entryId}/takes/${takeId}`), {
    method: "DELETE",
  });

  if (!response.ok) {
    await parseJsonResponse(response);
  }
}

export async function transcribeAudioJournalTake(entryId: number, takeId: number) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/${entryId}/takes/${takeId}/transcribe`), {
    method: "POST",
  });

  return parseJsonResponse<AudioJournalUploadResponse>(response);
}

export async function analyzeAudioJournalTake(entryId: number, takeId: number) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/${entryId}/takes/${takeId}/analyze-quality`), {
    method: "POST",
  });

  return parseJsonResponse<AudioJournalTakeRecord>(response);
}

export async function setActiveAudioJournalTake(entryId: number, takeId: number) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/${entryId}/takes/${takeId}/set-active`), {
    method: "POST",
  });

  return parseJsonResponse<AudioJournalTakeRecord>(response);
}

export async function setAudioJournalTrainingCandidate(
  entryId: number,
  takeId: number,
  payload: AudioJournalTrainingCandidatePayload,
) {
  const response = await fetch(await buildApiUrl(`/api/audio-journal/${entryId}/takes/${takeId}/training-candidate`), {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  return parseJsonResponse<AudioJournalTakeRecord>(response);
}

export async function buildAudioJournalTakeAudioUrl(entryId: number, takeId: number) {
  return buildApiUrl(`/api/audio-journal/${entryId}/takes/${takeId}/audio`);
}

File: backend/app/schemas/audio_journal.py
from datetime import datetime

from pydantic import BaseModel, Field


class AudioQualityMetrics(BaseModel):
    duration_seconds: float | None = None
    sample_rate: int | None = None
    channels: int | None = None
    file_format: str | None = None
    quality_status: str = "unknown"
    quality_score: float | None = None
    quality_summary: str | None = None
    quality_reasons_json: str | None = None
    noise_floor_db: float | None = None
    rms_db: float | None = None
    peak_db: float | None = None
    clipping_detected: bool = False
    silence_ratio: float | None = None
    snr_estimate_db: float | None = None


class AudioJournalRecordingAtmosphere(BaseModel):
    captured_at: datetime
    entry_id: int
    take_id: int
    take_number: int
    audio_filename: str
    duration_seconds: float | None = None
    sample_rate: int | None = None
    channels: int | None = None
    file_format: str | None = None
    quality_score: float | None = None
    noise_floor_db: float | None = None
    rms_db: float | None = None
    peak_db: float | None = None
    silence_ratio: float | None = None
    snr_estimate_db: float | None = None


class AudioJournalEntryCreate(BaseModel):
    title: str | None = Field(default=None, max_length=200)
    journal_date: datetime | None = None
    script_text: str | None = None
    notes: str | None = None
    tags_json: str | None = None
    voice_style: str | None = Field(default=None, max_length=100)
    metadata_json: str | None = None


class AudioJournalEntryUpdate(BaseModel):
    title: str | None = Field(default=None, max_length=200)
    journal_date: datetime | None = None
    script_text: str | None = None
    original_transcript_text: str | None = None
    notes: str | None = None
    tags_json: str | None = None
    voice_style: str | None = Field(default=None, max_length=100)
    active_take_id: int | None = None
    selected_training_take_id: int | None = None
    overall_quality_status: str | None = Field(default=None, max_length=32)
    metadata_json: str | None = None


class AudioJournalTakeCreate(BaseModel):
    take_type: str = Field(default="original", max_length=32)
    transcript_text: str | None = None
    transcript_source: str = Field(default="unknown", max_length=32)
    metadata_json: str | None = None


class AudioJournalTakeUpdate(BaseModel):
    transcript_text: str | None = None
    transcript_source: str | None = Field(default=None, max_length=32)
    transcription_status: str | None = Field(default=None, max_length=32)
    transcription_engine: str | None = Field(default=None, max_length=100)
    transcription_model: str | None = Field(default=None, max_length=255)
    quality_status: str | None = Field(default=None, max_length=32)
    quality_score: float | None = None
    quality_summary: str | None = None
    quality_reasons_json: str | None = None
    is_training_candidate: bool | None = None
    training_quality: str | None = Field(default=None, max_length=32)
    script_match_score: float | None = None
    metadata_json: str | None = None


class AudioJournalTrainingCandidateUpdate(BaseModel):
    is_training_candidate: bool
    manual_override: bool = False
    reason: str | None = None


class AudioJournalTakeRead(BaseModel):
    id: int
    entry_id: int
    take_number: int
    take_type: str
    created_at: datetime
    audio_path: str
    audio_filename: str
    transcript_text: str | None
    transcript_source: str
    transcription_status: str
    transcription_engine: str | None
    transcription_model: str | None
    duration_seconds: float | None
    sample_rate: int | None
    channels: int | None
    file_format: str | None
    quality_status: str
    quality_score: float | None
    quality_summary: str | None
    quality_reasons_json: str | None
    noise_floor_db: float | None
    rms_db: float | None
    peak_db: float | None
    clipping_detected: bool
    silence_ratio: float | None
    snr_estimate_db: float | None
    is_active: bool
    is_training_candidate: bool
    training_quality: str | None
    script_match_score: float | None
    metadata_json: str | None

    model_config = {"from_attributes": True}


class AudioJournalEntryRead(BaseModel):
    id: int
    title: str
    created_at: datetime
    updated_at: datetime
    journal_date: datetime
    script_text: str | None
    original_transcript_text: str | None
    notes: str | None
    tags_json: str | None
    voice_style: str | None
    active_take_id: int | None
    selected_training_take_id: int | None
    overall_quality_status: str
    metadata_json: str | None
    takes: list[AudioJournalTakeRead] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class AudioJournalUploadResponse(BaseModel):
    entry: AudioJournalEntryRead
    take: AudioJournalTakeRead


class AudioJournalListResponse(BaseModel):
    items: list[AudioJournalEntryRead]
